accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+* HttpClientHandler acts as a wrapper in most SDKs and the code would have to be duplicated
+* Platform specific implementation would need to write their own copy of proxy logic
+
+**Conclusion**
+
+This is very light-weight solution which pushes a lot of responsibility to optional components and build system. The possibility of including RID specific version of System.Net.Http.Private.dll inside Xamarin.iOS.nupkg needs to be confirmed.
+
+### Proposal #4: Add HttpClientHandler type forwarder to SDKs assemblies
+
+.NET builds ref version System.Net.Http for all optional components inside dotnet/runtime by using references to copies of optional component reference assemblies. The implementation then type forwards to optional components implementation assemblies.
+
+We could also use have only shared ref version of HttpClientHandler but then every optional SDKs would have to create their own copy of proxy logic as described in **Proposal #2**.
+
+**Located in dotnet/runtime repo**
+```c#
+// System.Net.Http.dll - ref for TFM net5.0-ios
+public class HttpClientHandler
+{
+  public HttpClientHandler ()
+  {
+  }
+
+  public long MaxInputInMemory { get; set; }
+
+  public NSUrlSessionHandlerTrustOverrideCallback TrustOverride { get; set; }
+  
+  // More unique or common properties to follow
+}
+```
+
+```c#
+// System.Net.Http.dll
+
+#if iOS
+// The reference for the type comes from Xamarin.iOS.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#elif (android)
+// The reference for the type comes from Xamarin.Android.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#else if ..
+
+#endif
+```
+
+**Disadvantages**
+
+* dotnet/runtime will need to include reference assemblies of all optional components to be able to build ref and impl assemblies.
+* Public HttpClientHandler API shape might be different between TFM as it’s controlled by optional components
+* Updating HttpClientHandler public surface for optional component will require build a new version of dotnet/runtime
+
+**Conclusion**
+
+The big advantage of this proposal is that the default HttpClientHandler would be have the most optimal implementation on every platform (performance and size wise). The availability of only platform specific properties on HttpClientHandler has its pros & cons as well.
+
+### Proposal #5: All HttpHandlers are fully implemented inbox
+
+Implement all platform-specific handlers fully inside dotnet/runtime using platform-specific interop libraries. We would extend low-level interop API which and use it to implement the managed API for HttpClientHandler and other types which are exposed from it.
+
+There is already a foundation for CF interop at [https://github.com/dotnet/runtime/tree/master/src/libraries/Common/src/Interop/OSX](https://github.com/dotnet/runtime/tree/master/src/libraries/Common/src/Interop/OSX). This will need to be extended to include Objective-C bridge for necessary APIs.
```

Comment:
It's not an extension to the current interop. What you call "CF interop" is C-based, with **manual** ref-counted memory, and p/invoke support (for C) are already there.

For ObjC you need things are much more complex - we have a runtime/tooling (the SDK) that just do that (the rest are just bindings, what we're talking to move here). Doing a submodule of `xamarin-macios` sounds counter-productive.

Reply:
Tweaked the text to make it clear the code would be independent to xamarin-macios